### ISSUE TITLE
INSTALL: Firedrake doesn't install on Python 3.12

### ISSUE THREAD

**Issue by JDBetteridge** [firedrakeproject/firedrake#3545]
**Describe the error**
In the title. This is an issue as Unbuntu 24.04 ships with Python3.12 by default.

**Steps to Reproduce**
`python3.12 ./firedrake-install`

**Expected behavior**
Firedrake installs.

**Error message**
Install script detects Python3.12 and exits

**Environment:**
 - OS: Linux (I'm guessing Mac too...)
 - Python version: 3.12

**Additional context**

**Comment by jmv2009** [comment]
```
Can not install Firedrake with Python 3.12 at the moment:
Some wheels are not yet available for Python 3.12 for some required package(s).
Please install with Python 3.11 (or an earlier version >= 3.8).
```
Hacking through the install scripts, it appears there is actually no problem, except for
a) numpy (version 1.24) (needed version restriction and remove from blacklist to go through)
b) mpi4py (needs python3-dev package on host system to build to go through)
Curiously, both are on the wheel_blacklist, which undermined the claimed reason 3.12 should not yet be used, as claimed.

**Comment by JDBetteridge** [comment]
I'm glad that hacking around has worked for you, but this won't solve the problem for all of our users. We still have some issues before we can claim that Firedrake is Python 3.12 ready. There are also good reasons to not use the numpy wheel related to BLAS/LAPACK ABI incompatibilities.

If you want to follow progress there is [a PR open](https://github.com/firedrakeproject/firedrake/pull/3546) which will address Python 3.12 installation when merged.

PS: I'm well aware that the message is two years old! I put it there :stuck_out_tongue:
